### PR TITLE
Langjian/cache miss

### PR DESCRIPTION
**Pull request by jianyinglang**
Add memory usage profiling function to compare the hardware memory usage and the cache-missed ng_function memory usage estimation.

### CODE REVIEW

File: src/ngraph_encapsulate_op.cc [hunk 1]
```diff
@@ -212,6 +213,26 @@ class NGraphEncapsulateOp : public OpKernel {
...
+
+    // the two fields we want
+    unsigned long vsize;
+    long rss;
+    {
+      std::string ignore;
+      std::ifstream ifs("/proc/self/stat", std::ios_base::in);
+      ifs >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore
+      >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore
+      >> ignore >> ignore >> vsize >> rss;
```
Comment (sayantan-nervana) [inline src/ngraph_encapsulate_op.cc]: this is a fun way of parsing the file :-D

We could use `ng.split` (https://github.com/NervanaSystems/ngraph/blob/master/src/ngraph/util.hpp#L79)

File: src/ngraph_encapsulate_op.cc [hunk 2]
```diff
@@ -262,11 +284,19 @@ class NGraphEncapsulateOp : public OpKernel {
...
+
       NGRAPH_VLOG(1) << "Compilation cache miss: " << ctx->op_kernel().name();
       OP_REQUIRES_OK(
           ctx, Builder::TranslateGraph(input_shapes, static_input_map, &m_graph,
                                        ng_function));
 
+      auto function_size = ng_function->get_graph_size()/1024.0; // kb unit
+
+      cout << "Resident memory = " << rss << endl;
+      cout << "ng_function measurement = " << ng_function->get_graph_size()/1024.0 << endl;
```
Comment (sayantan-nervana) [inline src/ngraph_encapsulate_op.cc]: This is `cout`. Should it be on `VLOG`?

File: src/ngraph_encapsulate_op.cc [hunk 3]
```diff
@@ -465,6 +509,8 @@ class NGraphEncapsulateOp : public OpKernel {
       try {
         op_backend->call(op_backend->compile(ng_function), ng_outputs,
                          ng_inputs);
+
+        //op_backend->remove_compiled_function(ng_function);
```
Comment (sayantan-nervana) [inline src/ngraph_encapsulate_op.cc]: This comment can be removed

Comment (jianyinglang) [reply src/ngraph_encapsulate_op.cc]: deleted it

File: src/ngraph_encapsulate_op.cc (hunk 2 above)
Comment (jianyinglang) [reply src/ngraph_encapsulate_op.cc]: Removed the cout

File: src/ngraph_encapsulate_op.cc (hunk 1 above)
Comment (jianyinglang) [reply src/ngraph_encapsulate_op.cc]: Fixed the way of file parsing

File: src/ngraph_encapsulate_op.cc [hunk 4]
```diff
@@ -284,6 +314,17 @@ class NGraphEncapsulateOp : public OpKernel {
...
+      mem_usage(vm, rss);
+      auto delta_vm_mem = vm - vm0;
+      auto delta_res_mem = rss - rss0;
+      NGRAPH_VLOG(1) << "NGTF_CACHE_PROFILE: " << ctx->op_kernel().name()
+                     << "  Step_ID: " << ctx->step_id();
+      NGRAPH_VLOG(1) << "       Delta Virtual Memory Measurment: "
+                     << delta_vm_mem
+                     << "  Delta Resident Memory Measurment: " << delta_res_mem
+                     << "  Function Memory Measurment: " << function_size;
+      NGRAPH_VLOG(1) << "       Resident Memory: " << rss;
```
Comment (shresthamalik) [inline src/ngraph_encapsulate_op.cc]: @avijit-nervana 
Should we use an environment variable for this like NGRAPH_TF_CACHE_LOGS=1 or NGRAPH_VLOG is ok?

**Comment by jianyinglang** [comment]
Close this PR and opened another one : https://github.com/NervanaSystems/ngraph-tf/pull/428

**Pull request closed without merging**